### DISCUSSION TITLE
multi-tenancy implementation with hasura-auth

### DISCUSSION THREAD

**Discussion by xmlking** [nhost/nhost#3016] [Ideas]
Creating this discussion on **multi-tenancy**,  to capture ideas from other nhost users.

Original  discussion on  **Discord**   https://discord.com/channels/552499021260914688/1198027502052970648

##  Requirements

1. System  should support **multi-tenancy** as described below

![image](https://github.com/user-attachments/assets/005f22c5-00bc-487c-ac9f-2a5b600d3d7d)

  - **John** is affiliated with two organizations, using the email `john@email.com` as his unique identifier. He holds the position of `org:admin` in `Organization A` and is a `org:guest` in `Organization B`.
  - **Sarah** is associated with a single organization and uses the email `sarah@email.com` as her unique identifier. She is the `org:admin`  of `Organization B`.

2. User should be able to switching organizations with-in App and able to interact with App that are allow for user's role with-in switched organization

<img width="787" alt="image" src="https://github.com/user-attachments/assets/112bec49-4d95-4e4b-ad56-84009d679b45">

Switching organization (and roles) should done via refreshing  **JWT** with new `org-id` from `hasura-auth`
  
How it could be done:
- App will to send switching org as  `x-hasura-org` header and current JWT as `Authorization: "Bearer ...`  header to **hasura-auth** refresh token API.
- **hasura-auth** validate value of this header via permission check if user is indeed a member of this Organization
- App will use new refreshed JWT token for subsequent GraphQL API calls. 

## Data Modeling

**public.user_org_roles** table 

<img width="1779" alt="image" src="https://github.com/user-attachments/assets/8514be5f-c5ca-4a60-8a52-7c8283d7e87b">

**public.organizations** table

<img width="1049" alt="image" src="https://github.com/user-attachments/assets/0b26cb31-654c-4ae9-90a4-c8760644f728">

**auth.user** table

<img width="414" alt="image" src="https://github.com/user-attachments/assets/203d929e-0a88-41d3-8332-5055fc36c6f1">

<img width="1067" alt="image" src="https://github.com/user-attachments/assets/c6f98355-5db8-4f5e-9f22-60bb3d29fdb3">

<img width="1067" alt="image" src="https://github.com/user-attachments/assets/ab6f3836-61a7-4c40-86e2-a61440ed44da">

## Appendix
Example JWT token

```json
{
  "exp": 1731810450,
  "https://hasura.io/jwt/claims": {
    "x-hasura-allowed-orgs": "{\"org1\",\"org2\"}",
    "x-hasura-allowed-roles": [
      "org:owner",
      "org:admin",
      "app:editor"
    ],
    "x-hasura-default-org": "org1",
    "x-hasura-default-role": "org:admin",
    "x-hasura-user-email": "sumo@demo.com",
    "x-hasura-user-id": "076a79f9-ed08-4e18-a5c3-8d4e0aa269a3",
    "x-hasura-user-is-anonymous": "false"
  },
  "iat": 1731806850,
  "iss": "hasura-auth",
  "sub": "076a79f9-ed08-4e18-a5c3-8d4e0aa269a3"
}
```

## Ref
- https://github.com/hasura/graphql-engine/issues/1601

**Comment by dbarrosop** [comment]
Hello,
thanks for the write-up, it helps us understand what people are doing. The main challenge here is making this generic as different users will model their application differently and also taking into account custom claims might be scoped by tenant. For instance, for an application like our nhost cloud we could do something:

```
    "x-hasura-organization": "org1",
    "x-hasura-permissions": [
       "project:*:settings:write",            # user can modify settings on all projects
       "project:*:run:*:write",               # user can manipulate all run services on all projects
       "billing:read",                        # user can see billing information
       "project:asdwqexzcsdf:database:write", # only allow access to the database on a given project
    ],
```

And, of course, switching organization should require refreshing the permissions claim. Keep in mind I am not trying to argue about the merits of one solution or another. Simply that different applications may want to model this differently so the solution we come up with needs to be completely application agnostic. 

Right now what I am thinking is two possible solutions:

1. The easiest one would be to introduce a webhook that auth could delegate the custom claims creation to. For instance:

user -> POST /token?organization=org1 (this is today's /token endpoint but accepting arbitrary parameters)
auth -> POST /webhook?organization=org1 (hasura-auth calls the webhook passing the authentication headers and the query parameters sent to /token. The webhook can then validate the user and return the claims that the auth service will simply attach to the token.

2. An alternative would be to come up with a more extensible mechanism to define custom claims. Something that allows for scoping based on some tenant key. For instance:

```toml
[auth.session.accessToken.multitenantKey]
# this can then be used as a variable in the jsonpath in custom claims
key = 'organization-id'        # this would be the custom claim
value = 'organizations[*].id'  # when a user requests a token for a given tenant the tenant key
                               # must be inside this query or it will error out

[[auth.session.accessToken.customClaims]]
key = 'allowed-organizations'
value = 'organizations[*].id'

[[auth.session.accessToken.customClaims]]
key = 'permissions'
value = 'organizations[$tenant].permissions'

[[auth.session.accessToken.customClaims]]
key = 'role'
value = 'organizations[$tenant].role'
```

Hopefully we could come up with something not very complex that works for most use-cases.

**Comment by xmlking** [comment]
I would agree that any solution that has minimal impact on the changes needed for hasura-auth would be ideal. 
- **Option 1** appears to be the more versatile option, but it does require multiple network hops, and this can negatively affect the app's reliability and availability.
- **Option 2**  enables native support multi-tenancy,  though, for my use case: database `permissions` are bundled under `roles` , I only needed `roles` switched every time the `org` is switched.  For this, I need to store user's _roles_ per _organization_  in special table.

For option-2, thinking of tenant switch workflow like:
1. `webapp` request org change to `hasura-auth` server by passing current JWT
2.  `hasura-auth` update `default_org`   column in `auth.users`  table after verifying  in database table `auth.user_org_roles` or in `x-hasura-allowed-orgs` claim.
3. `hasura-auth`  then generare fresh JWT with requested `default-org/default-role` claims change and return JWT
4. `webapp`  repeat above steps every time it needs org switch  

**Impact:** This need new column `default_org` added to  `auth.users` table and new `auth.user_org_roles` table to lookup user's org specific `allowed-roles` and   org specific `default-role`

**Reply by dbarrosop** [reply]
> hasura-auth update default_org column in auth.users table after verifying in database table auth.user_org_roles or in x-hasura-allowed-orgs claim.

Why do you need this? This is something hasura-auth will most certainly not do.

> Impact: This need new column default_org added to  auth.users table and new auth.user_org_roles table to lookup user's org specific allowed-roles and org specific default-role

If your particular implementation needs this, this should be in your application schema, not the auth one. As mentioned before, whatever change we do needs to be application agnostic.

**Reply by xmlking** [reply]
When user singin, the App needs to know  current default_org, so that it can fetch data related to that org.
The auth server need to retrieve current org form some persistent state, and pass it  to app as JWT claim. 

I though auth.user (metadata) table might be the best please to store current default_org info.

**Reply by dbarrosop** [reply]
> the App needs to know current default_org, so that it can fetch data related to that org

Not necessarily, you can read the claim from the JWT token. So basically this is what I'd expect people to do:

1. refresh token /token?organization_id=xxxxxxxx // this is the multitenant key I mentioned before
2. auth adds the claim `X-Hasura-Organization-Id: xxxxxxxxx` to the JWT token alongside the rest of the claims scoped for organization xxxxxxxx
3. app reads reads the organization-id claim to understand the context
4. when user wants to switch orgs refresh the token again /token?organization_id=yyyyyyyyy
5. auth now adds the claim `X-Hasura-Organization-Id: yyyyyyyyyy` to the JWT token alongside the rest of the claims scoped for organization yyyyyyyyy

With this flow there is no need to modify/read anything from the database, everything resides in the access token which means a few things:
- should be completely application agnostic (only requisite is to have a single multitenant key)
- no need to modify anything on the database
- it should work in hasura and also in any other custom backend

**Reply by xmlking** [reply]
Have few more questions 
How does the Application know xxxxxxxx value to make the first call  to /token?organization_id=xxxxxxxx ?
If we go with this approach , will nhost team implement /token?organization_id endpoint ?

**Reply by dbarrosop** [reply]
> How does the Application know xxxxxxxx value to make the first call to /token?organization_id=xxxxxxxx ?

Yes, that would be the current `/token` endpoint with an optional parameter that should:

- match the multitenantKey.Key defined above (maybe even hardcode to `tenant` and then map it internally in the token to the user-defined key)
- have a value that is included in the response of the query `multitenantKey.value`, otherwise return an error
- default to the first returned value in the query if none specified

**Comment by xmlking** [comment]
Exploring ideas from @viktorfa 
https://github.com/viktorfa/nhost-saas-boilerplate/tree/main

**Reply by xmlking** [reply]
@dbarrosop I think it provide flexibility for Devs if `hasura-auth` allow  overwrite   [`x-hasura-allowed-roles`] and [`x-hasura-default-role`] claims  like other custom claims.
https://github.com/nhost/hasura-auth/blob/3e2ffd373eb32403f66d66bbde5c2d54a778616d/src/utils/jwt/generate.ts#L77  

I like to use **org** specific `allowed-roles` and  `default-role` and  taking advantage of  [Inherited Roles](https://hasura.io/docs/2.0/auth/authorization/inherited-roles/) while defining  permissions in hasura. 
 
 
 with that I will have flexibility to add _custom claims_ in JWT. e.g., 
 
 ```
x-hasura-allowed-roles  = user roles (in auth.user_roles) + org role (in public.org_user) 
x-hasura-default-role =  user_profile.org_user.role    i.e.,  org role (in public.org_user)
x-hasura-org-ids = org_users.org_id[]
x-hasura-org-id = user_profile.org_user.org_id
x-hasura-org-user-id = user_profile.current_org_user
```

**Reply by dbarrosop** [reply]
would you mind describing the use case for this feature? Specifically, why is it needed. Those two claims are used to map a user to a hasura role, unless you have one (or many) hasura roles per organization adding the organization roles to those claims serve no real purpose. And even if you have different hasura roles per organizations you could simply add them to the user_roles. So in summary:

1. does hasura need to know about these extra roles to map the user to a \*\*hasura role\*\*?
2. if yes to (1), why can't it be then in user_roles table?
3. if no to (1), why do you need it here instead of in a custom claim like `x-hasura-organization-roles`?

**Reply by xmlking** [reply]
#### How my org _roles_ are arranged:
1. Org roles are used for _Admin UI WebApp_  users only.
2. Org roles are mutually exclusive for an user within an organization. i.e., **John** cannot have both **org:admin** and **org:member** role in  **Organization A**.
3. Org roles are **hierarchical**. i.e., `org:owner` > `org:admin`  > `org:member`  etc. 

> _System roles (user, me)_ are not relevant w.r.t _Admin UI WebApp_ , but I could use them for  permissioning `CLI App(agent)` users  in the future. (CLI App also access some hasura data via GQL API)


This is what I have in my `nhost/metadata/inherited_roles.yaml` file
```yaml
- role_name: me
  role_set:
    - user
    - private
- role_name: org:admin
  role_set:
    - org:member
- role_name: org:billing
  role_set:
    - user
- role_name: org:member
  role_set:
    - user
- role_name: org:owner
  role_set:
    - org:admin
    - org:billing
- role_name: user
  role_set:
    - public
```

#### How I want org switching happen:

When **John** _SignIn_ and access  **Organization A** (where his default org-role is  **org:admin**) , we like to issue _AccessToken (AT)_ with following claims:

```json
{
  "https://hasura.io/jwt/claims": {
    "x-hasura-Org-Id": "4f073c66-e9e5-4f6c-a425-a48ed6fd1e04",
    "x-hasura-Org-Ids": "{\"4f073c66-e9e5-4f6c-a425-a48ed6fd1e04\",\"406a6fc6-ecb9-4f0f-9812-8a2abb717520\"}",
    "x-hasura-Org-User-Id": "f53dfbf7-bf04-4b5e-9adf-1e3ce7b3d514",
    "x-hasura-user-email": "john@email.com",
    "x-hasura-allowed-roles": [
      "me",
      "user",
      "org:admin"
    ],
    "x-hasura-default-role": "org:admin",
    "x-hasura-user-id": "193c2cac-6a22-4e9c-b5f8-93fe464c9875",
    "x-hasura-user-is-anonymous": "false"
  }
```
Then  **John**  _switch_ to   **Organization B** (where has default org-role is  **org:member**)  via UI, then **AT** get refreshed, we issue new  _AccessToken_ with following claims:


```json
{
  "https://hasura.io/jwt/claims": {
    "x-hasura-Org-Id": "406a6fc6-ecb9-4f0f-9812-8a2abb717520",
    "x-hasura-Org-Ids": "{\"4f073c66-e9e5-4f6c-a425-a48ed6fd1e04\",\"406a6fc6-ecb9-4f0f-9812-8a2abb717520\"}",
    "x-hasura-Org-User-Id": "b7900dd8-f06e-4907-9fbd-16f3cb0386e7",
    "x-hasura-user-email": "john@email.com",
    "x-hasura-allowed-roles": [
      "me",
      "user",
      "org:member"
    ],
    "x-hasura-default-role": "org:member",
    "x-hasura-user-id": "193c2cac-6a22-4e9c-b5f8-93fe464c9875",
    "x-hasura-user-is-anonymous": "false"
  }
```

### For example,  I will be using the org `roles` to define hasura permissions for a given table e.g: _Projects_:
1. `org:member` can SELECT rows belongs to his `x-hasura-Org-Id` org, but cannot DELETE any rows.
2. `org:admin` can SELECT rows belongs to his `x-hasura-Org-Id` org due to role _inheritance_, and also  UPDATE those row. 
3.  `org:owner` can SELECT rows belongs to his `x-hasura-Org-Id` org due to role _inheritance_, and also  UPDATE/DELETE those row. 
6.  `org:billing`  can neither  SELECT nor  UPDATE/DELETE  rows belongs to his `x-hasura-Org-Id` org.
7. All users ( org:member, org:admin, org:owner, org:billing, user) can SELECT/UPDATE their personal profile data in `auth.users` table. This permission can be granted at `user` role level and apply to all higher roles via _inheritance_

> Note: all my application tables has `created_by`, `created_at`, `updated_by`, `updated_at` and `org_id`  columns and used to check  permissions.

To answer your questions:
> does hasura need to know about these extra roles to map the user to a **hasura role**? 

**YES**

> if yes to (1), why can't it be then in user_roles table?

`auth.user_roles` table cannot maintain user roles at organization granular level. 

> if no to (1), why do you need it here instead of in a custom claim like x-hasura-organization-roles?

as stated above, I am using org role to define hasura table permissions

**Reply by xmlking** [reply]
An example of complex org role permissions: 

| Action                                                        | Billing | Member | Admin | Manager | Owner |
| ------------------------------------------------------------- | ------- | ------ | ----- | ------- | ----- |
| Can see and edit billing information and subscription details | ✓       |        |       |         | ✓     |
| Can see and edit legal and compliance details                 | ✓       |        |       |         | ✓     |
| Can view and act on issues (such as assigning and resolving)  |         | ✓      | ✓     | ✓       | ✓     |
| Can join and leave existing teams                             |         | ✓      | ✓     | ✓       | ✓     |
| Can add repositories (through GitHub integrations)            |         | ✓      | ✓     | ✓       | ✓     |
| Can create and remove teams from the organization             |         |        | ✓     | ✓       | ✓     |
| Can change project settings                                   |         | ✓*     | ✓     | ✓       | ✓     |
| Can create and remove projects for the organization           |         | ✓**    | ✓     | ✓       | ✓     |
| Can edit Global Integrations                                  |         |        | ✓     | ✓       | ✓     |
| Can remove repositories                                       |         |        | ✓     | ✓       | ✓     |
| Can add, remove, and change Org Members                       |         |        |       | ✓       | ✓     |
| Can change organization settings, including open membership   |         |        |       | ✓       | ✓     |
| Can assign Team Admin role                                    |         |        |       | ✓       | ✓     |
| Can add, remove, and edit Team Contributors                   |         |        |       | ✓       | ✓     |
| Can add projects for teams                                    |         |        |       | ✓       | ✓     |
| Can remove projects from teams                                |         |        |       | ✓       | ✓     |
| Can transfer projects to another organization                 |         |        |       |         | ✓     |
| Can remove an organization                                    |         |        |       |         | ✓     |

**Reply by dbarrosop** [reply]
Ok, thanks a lot for the added details. This actually makes a lot of sense. It makes so much sense that it might be worth integrating directly into auth. Something like the following:

1. Add a `tenants` table with just `id`, `created_at`, `updated_at` and `metadata` columns
2. Add a `user_roles_tenants` table with `user_id`, `tenant_id`, `is_default_role` columns
3. Add a setting in the Auth service to enable/disable multitenancy
4. If multitenancy is disabled, things work like they do today
5. If multitenancy is enabled:
   - the `/token` endpoint accepts an extra argument `tenant_id` to populate `x-hasura-tenant-id` claim + select the appropiate roles for the tenant (see next point).
   - we use the `user_roles_tenants` table to fill the allowed roles and default roles instead of the `user_roles` table

I think this would be generic enough to implement what you are suggesting without imposing too much to users.

If this sounds reasonable we could try experimenting and testing a bit, this should also be much simpler to implement than what I was proposing.

**Reply by xmlking** [reply]
I am very excited to see your proposal  for multi-tenancy  implemented  natively in **hasura-auth**. This should have no side-effects on current single-tenant users. 

Please note `tenants` table should have some extensibility  to support some business  specific metadata, in my case, I have `auto_enroll` column to automatically add  some users with specific email domain to an org when users registers with such email.   
> for reference please see my experimental repo with schema [here](https://github.com/xmlking/nhost-multi-tenancy-experiment)

One challenge with `is_default_role` column would be, **hasura-auth** process has to reset other  user org role's `is_default_role` column to `false` whenever org default role changed. I was using a trigger for that:

```sql
CREATE FUNCTION public.enforce_single_default_role() RETURNS trigger
    LANGUAGE plpgsql
    AS $$
BEGIN
  -- Update all other rows to set is_default_role to false
UPDATE public.user_org_roles
SET is_default_role = false
WHERE user_id = NEW.user_id
  AND organization = NEW.organization
  AND is_default_role = true
  AND id <> NEW.id;
RETURN NEW;
END;
$$;

CREATE TRIGGER trg_enforce_single_default_role BEFORE INSERT OR UPDATE OF is_default_role ON public.user_org_roles FOR EACH ROW WHEN ((new.is_default_role = true)) EXECUTE FUNCTION public.enforce_single_default_role();
COMMENT ON TRIGGER trg_enforce_single_default_role ON public.user_org_roles IS 'trigger to set value of column "is_default_role" to "false" for all other rows, on a row insert or update';
```

**Reply by dbarrosop** [reply]
> One challenge with is_default_role column would be...

Yeah, I was expecting something like that. Nothing that we haven't done before so not super worried :)

> Please note tenants table should have some extensibility...

Yeah, in your particular case my proposal for you would be something like:

1. create table `public.tenants_auto_enroll` with columns `tenant_id` and `email_domain`.
2. add a trigger on `auth.users` inserts that extracts the email domain, looks for it in the table `public.tenants_auto_enroll` and inserts any relevant entries into `auth.user_roles_tenant`

Point being, I think you should be able to add this on top of what auth provides without auth actually having to know anything about it.

**Comment by xmlking** [comment]
I am experimenting with other minimalistic approach to achieve multi-tenancy with current `hasura-auth`:

This also have advantage to support  **partial-multi-tenancy** in a single Application , i.e., Regular Users (`user`, `me`) stay single-tenant as before, and Org Users(`user`, `me`, `org:member`) can switch between multiple Orgs carrying org's defaultRole.

### Limitations
1. Org Users will have only one active Org role at a time. -- `that is acceptable for my case` 
2. Need a [patch](https://github.com/nhost/hasura-auth/pull/594) for `hasura-auth`


sample implementation : 
https://github.com/xmlking/nhost-multi-tenancy-experiment
